### DISCUSSION TITLE
Events support!

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#225] [General]
For those of you that might've missed todays update change log:  twinBASIC v0.10.1841 now supports Events!

https://marketplace.visualstudio.com/items/twinbasic.twinbasic/changelog

In typical twinBASIC style, we've also enhanced them to remove some of the VBA/VB6 restrictions by allowing events to have optional values, ParamArrays, and customizable DispIDs (using the DispId attribute).

Have fun!
Wayne

**Comment by bclothier** [comment]
Fantastic news! 

As a clarification - we are still awaiting support for explicitly defining & assigning source interfaces, correct? At the present the support is basically the same thing we have in VB* where it is automatically generated behind the scene, right?

**Comment by WaynePhillipsEA** [comment]
@bclothier yes, that's right.   Could you give me a rough example of the syntax you'd be wanting for defining them more explicitly?  You can open an issue for it if you'd prefer.

**Reply by bclothier** [reply]
In C# I would have done something like this:
```
[
  ComVisible(true),
  Guid("..."),
  InterfaceType(ComInterfaceType.InterfaceIsDual)
]
public Interface IFoo
{
  void DoFoo();
}

[
  ComVisible(true),
  Guid("...")
  InterfaceType(ComInterfaceType.InterfaceIsIDispatch)
]
public interface IFooEvents
{
  void FooHappened();
}

[
  ComVisible(true),
  Guid("..."),
  ProgId("..."),
  ClassInterface(ClassInterfaceType.None),
  ComDefaultInterface(typeof(IFoo)),
  ComSourceInterfaces(typeof(IFooEvents))
]
public class Foo : IFoo
{
  public delegate void FooHappenedDelegate();
  public event FooHappenedDelegate FooHappened;

  public void DoFoo()
  {
      FooHappened?.Invoke();
  }
}
```

To make it more BASIC, it probably would be something like this (using minimal code only). The attributes can be added for more control over Guids, etc. Unsure if it's required that the source interface be a `IDispatch` rather than dual.

```
[
  InterfaceType(ComInterfaceType.InterfaceIsIDispatch)
]
Public Interface IFooEvents
  Sub FooHappened()
End Interface

Public Class Foo
  Implements Events IFooEvents

  Public Sub Foo()
    RaiseEvents IFooEvents.FooHappened()
  End Sub
End Class
```

If we choose to extend the keyword `Implements` rather than using attributes, I can see those variants:

```
Implements Default IFoo
Implements IBar 'Non-default interface
Implements Default Events IFooEvents
Implements Events IBarEvents 'Non-default event interface
```

which corresponds to the MIDL's attributes for the `source` and `default` that can be decorated on interfaces within the `coclass` block. The alternative is to use attributes to decorate what is default:

```
[
  ComDefaultInterface(TypeOf IFoo)
  ComDefaultSourceInterface(TypeOf IFooEvents)
]
Public Class Foo
  Implements IFoo
  Implements Events IFooEvents
  Implements IBar
  Implements Events IBarEvents

...
End Class
```

I am unsure if it is possible / desirable to use attributes solely for defining the source interfaces.
```
[
  ComEventInterface(TypeOf IFooEvents)
]
Public Class Foo
  Public Sub
    RaiseEvents IFooEvents.FooHappened() 
  End Sub
End Class
```

I hope that helps give you an idea of how we can gain more fine-tuning over the COM design around the source interfaces.

**Reply by bclothier** [reply]
One more alternative that is arguably more BASIC-like:
```
Public Interface Foo
  Event Sub FooHappened()
  Sub DoFoo()
End Interface
```

Currently, the VB* does not allow you to `Implements` a class and be able to `RaiseEvents` events defined on the implemented class. That restriction would be relaxed. However, the downside is that this is still managed by the compiler behind the scene and doesn't help if you need to implement an external source interface.

**Reply by wqweto** [reply]
I'm not sure you can have more that one source dispinterface per coclass (what the class events compile to in the typelib) so not sure if this is possible to raise some other coclass source dispinterface events.

cheers,
&lt;/wqw>

**Reply by bclothier** [reply]
From `MSACC.OLB` for `Form` coclass:
```
[
  uuid(7398AAFD-6527-48C7-95B7-BEABACD1CA3F),
  helpcontext(0x00003576)
]
coclass Form {
    [default] interface _Form3;
    [source] interface _FormEvents;
    [default, source] dispinterface _FormEvents2;
};
```

Yes it's legal. You can only have one default interface and one default source interface but you can have as many interfaces (source or not) as you want for a coclass.

**Reply by wqweto** [reply]
I mean source **disp** interface as VBA/VB6 cannot listen for events from non-disp source interfaces i.e. `[source] interface _FormEvents` cannot be listened upon in VBA/VB6 using `WithEvents`.

**Comment by wqweto** [comment]
This is great news!

It seems my [cAsyncSocket](https://github.com/wqweto/VbAsyncSocket/blob/master/src/cAsyncSocket.cls) implementation is working now mostly without modifications, including the ASM thunks for the message queue which frankly surprised me a little :-))

Edit: Wow, this even allows [cVncServer](https://github.com/wqweto/VbVncServer/blob/master/src/cVncServer.cls) to compile and run under TB which is very impressive IMO, considering the project is using an external DirectX 11 typelib too.

Btw, do you plan on allowing for the referenced typelibs to be embedded in the workspace so that the projects can be distributed in single file without external typelib dependencies for the source code? That would be handy.

**Reply by WaynePhillipsEA** [reply]
@wqweto That's really awesome!!  

If the type libraries are in separate TLB files, you should be able to drop them into the Resources -> TypeLibraries folder, and then use the 'Load From File...' option underneath the project references in the settings to select it.  However, I must admit that I haven't used that feature for quite some time so I can't guarantee that it's working at the moment.

**Reply by wqweto** [reply]
Btw, completing source events is like 50% ready (starting w/ initial public release) for big projects trial migration. I mean as big as Olaf's RC6 for instance.

Rest of the wrinkles can be solved by precise find&replace edits or minor source code modifications (e.g. existing `GoTo Continue` statement will not compile as reserved word `Continue` cannot be a name for a label in TB anymore).

In this regard new issues downpour is yet to come when large projects trial migrations get going and some obscure edge cases are hit :-))

**Reply by WaynePhillipsEA** [reply]
Bring on the edge cases! 😎

Things like the Continue statement will be disabled with Legacy (#65) mode, to ease migration.  There aren't masses amounts of new syntax that will clash, so Legacy mode won't be needed in many cases.
